packages/SwingSet/src/liveslots/virtualReferences.js

```diff
@@ -160,6 +218,46 @@ export function makeVirtualReferenceManager(
     kindInfoTable.set(`${kindID}`, { reanimator, deleter, durable });
   }
 
+  /**
+   * Compare two arrays (shallowly) for equality.
+   *
+   * @template T
+   * @param {T[]} a1
+   * @param {T[]} a2
+   * @returns {boolean}
+   */
+  const arrayEquals = (a1, a2) => {
+    assert(Array.isArray(a1));
+    assert(Array.isArray(a2));
+    if (a1.length !== a2.length) {
+      return false;
+    }
+    return a1.every((elem, idx) => Object.is(a2[idx], elem));
+  };
+
+  /**
+   * Check a list of facet names against what's already been established for a
+   * kind.  If they don't match, it's an error.  If nothing has been established
+   * yet, establish it now.
+   *
+   * @param {string} kindID  The kind we're talking about
+   * @param {string[]|null} facetNames  A sorted array of facet names to be
+   *    checked or acquired, or null if the kind is unfaceted
+   */
+  function checkOrAcquireFacetNames(kindID, facetNames) {
+    const kindInfo = kindInfoTable.get(`${kindID}`);
+    assert(kindInfo, `no kind info for ${kindID}`);
+    if (kindInfo.facetNames !== undefined) {
+      if (facetNames === null) {
+        assert(kindInfo.facetNames === null);
+      } else {
+        assert(arrayEquals(facetNames, kindInfo.facetNames));
```

Comment:
This `assert` should have a descriptive string with something like `all virtual objects of the same Kind must have the same facet names`, otherwise userspace will get a pretty opaque error. If there's an easy way to include the name/tag of the Kind, and/or the list of expected/received facet names, that will make their life even easier.